inject/src/main/java/io/micronaut/inject/beans/visitor/BeanPropertyWriter.java

```diff
@@ -155,22 +167,165 @@ public void accept(ClassWriterOutputVisitor classWriterOutputVisitor) throws IOE
 
             if (readOnly) {
                 // override isReadOnly method
-                final GeneratorAdapter isReadOnly = startPublicMethodZeroArgs(classWriter, boolean.class, "isReadOnly");
+                final GeneratorAdapter isReadOnly = startPublicFinalMethodZeroArgs(classWriter, boolean.class, "isReadOnly");
                 isReadOnly.push(true);
                 isReadOnly.returnValue();
                 isReadOnly.visitMaxs(1, 1);
                 isReadOnly.endMethod();
             }
 
             if (writeMethod != null && readMethod == null) {
-                // override isReadOnly method
-                final GeneratorAdapter isWriteOnly = startPublicMethodZeroArgs(classWriter, boolean.class, "isWriteOnly");
+                // override isWriteOnly method
+                final GeneratorAdapter isWriteOnly = startPublicFinalMethodZeroArgs(classWriter, boolean.class, "isWriteOnly");
                 isWriteOnly.push(true);
                 isWriteOnly.returnValue();
                 isWriteOnly.visitMaxs(1, 1);
                 isWriteOnly.endMethod();
             }
 
+            // override isMutable method
+            boolean isMutable = this.isMutable;
```

Comment:
Like above It would be nice to skip modification if the value is the same.
Maybe it can even delegate to an existing method to modify immutable objects like generated by https://projectlombok.org/features/Value

```java
ValueExample withAge(int age) {
    return this.age == age ? this : new ValueExample(name, age, score, tags);
  }
```

Reply:
I think delegating to existing methods is a nice idea and may be possible to implement once #4813 is merged as we can use the new `ElementQuery` API to find the methods